### PR TITLE
adds batch jobs for long-running geometry creation

### PR DESCRIPTION
**Pull request by andy-esch**
For larger tables written to carto with the `lnglat` flag, the geometry creation step times out. To get around this, we can send the job to the [SQL Batch API](https://carto.com/docs/carto-engine/sql-api/batch-queries/). The upside is that there will not be a timeout, the downside is that it has to sit in a queue.

The new flow for `cc.write` is that if the dataframe has more than 1e5 rows, the geometry creation query will be sent to the Batch SQL API and a `BatchJobStatus` object will be returned. This object can be polled with the `.status()` method. If the number of rows is at or below 1e5, the normal SQL API will be used for geometry creation and `cc.write` will return `None`.

Example usage:

```python
import time
job = cc.write(df, 'new_table',
               lnglat=('lng_col', 'lat_col'))
while job is not None:
    curr_status = job.status()['status']
    if curr_status in ('done', 'failed', 'canceled', 'unknown', ):
        print(curr_status)
        break
    time.sleep(5)
```

ToDos

- [x] turn the `pandas.groupby` into a generator
- [x] ensure that there's test coverage to all the new pieces
  - [x] `BatchJobStatus`
  - [x] lnglat for batch jobs

### CODE REVIEW

**Comment by andy-esch** [comment]
@dmed256, could you take a look at this?

File: cartoframes/context.py [hunk 1]
```diff
@@ -175,20 +178,30 @@ def write(self, df, table_name, temp_dir='/tmp', overwrite=False,
...
-                       lng=lnglat[0],
-                       lat=lnglat[1]))
+            query = '''
+                    UPDATE "{table_name}"
+                    SET the_geom = CDB_LatLng("{lat}"::numeric,
+                                              "{lng}"::numeric)
+                    '''.format(table_name=final_table_name,
+                               lng=lnglat[0],
+                               lat=lnglat[1])
+            if df.shape[0] > 100000:
```
Comment (dmed256) [inline cartoframes/context.py]: I think it should be:
```if df.shape[0] > MAX_IMPORT_ROWS:```

File: cartoframes/context.py [hunk 2]
```diff
@@ -175,20 +178,30 @@ def write(self, df, table_name, temp_dir='/tmp', overwrite=False,
...
+                batch_client = BatchSQLClient(self.auth_client)
+                status = batch_client.create([query, ])
+                tqdm.write('Table successfully written to CARTO: '
+                           '{base_url}dataset/{table_name} . `the_geom` '
```
Comment (dmed256) [inline cartoframes/context.py]: To make sure the `/` are there:
```
os.path.join(base_url, 'dataset', table_name)
```

File: test/test_context.py [hunk 3]
```diff
@@ -189,6 +195,16 @@ def test_cartocontext_write(self):
         self.assertEqual(resp['rows'][0]['num_rows'],
                          resp['rows'][0]['num_geoms'])
 
+        # test batch lnglat behavior
+        n_rows = 100001
```
Comment (dmed256) [inline test/test_context.py]:
```
n_rows = context.MAX_IMPORT_ROWS + 1
```

File: cartoframes/context.py (hunk 2 above)
Comment (dmed256) [reply cartoframes/context.py]: Maybe it's ok the way you have it, it's in other places in the code

Comment (andy-esch) [reply cartoframes/context.py]: Might as well be careful

File: cartoframes/context.py [hunk 4]
```diff
@@ -1105,3 +1120,47 @@ def _df2pg_schema(dataframe, pgcolnames):
...
+        self.job_id = job['job_id']
+        self.last_status = job['status']
+        self.batch_client = batch_client
+
+    def __repr__(self):
+        return ('BatchJobStatus(job_id=\'{job_id}\', '
+                'last_status=\'{status}\')'.format(job_id=self.job_id,
+                                                   status=self.last_status))
+
+    def status(self):
```
Comment (andy-esch) [inline cartoframes/context.py]: @dmed256 you suggested in https://github.com/CartoDB/cartoframes/issues/85 a method for returning dataframes from longer running queries. Here we could have a flag or another method for returning the lnglat'd version of the table/dataframe. I envision us adding more than the lnglat flag for adding geometries.. e.g., street-level geocoding, routing, etc. 

But here, we could have a simple new method (or flag to status), which indicates that when `status=done`, do `cc.read('tablename')` to return the carto'd dataframe.

What do you think? Too soon?

Comment (andy-esch) [reply cartoframes/context.py]: I think too soon. I want to scope it out more before including anything else here.

**Comment by andy-esch** [comment]
To avoid the issues discussed in #153 and #210, I've done some refactoring to investigate more efficient ways of getting the tables to (1) display in the dashboard and get a dataset page without being a ghost table, and (2) avoid timeouts by finding cleverer ways of populating the geometry field by avoiding `UPDATE` statements and offloading the queries to the Batch SQL API.

Some timings without the Batch SQL API for the entire write process:
screen shot 2017-09-14 at 12 51 44

**Pull request merged**